drivers/sensor/ti/ina700/ina700.c

```diff
@@ -0,0 +1,624 @@
+
+		val->val1 = temperature / 1000000L;
+		val->val2 = temperature % 1000000L;
+	} else if (chan == SENSOR_CHAN_CURRENT) {
+		current = data->current * INA700_CURRENT_UA_LSB;
+
+		val->val1 = current / 1000000L;
+		val->val2 = current % 1000000L;
+	} else if (chan == SENSOR_CHAN_POWER) {
+		power = data->power * INA700_POWER_UW_LSB;
+
+		val->val1 = power / 1000000U;
+		val->val2 = power % 1000000U;
+	} else if (ina_chan == SENSOR_CHAN_ENERGY) {
+		energy = data->energy * INA700_ENERGY_UJ_LSB;
+
+		val->val1 = energy / 1000000U;
+		val->val2 = energy % 1000000U;
+	} else if (ina_chan == SENSOR_CHAN_CHARGE) {
+		charge = data->charge * INA700_CHARGE_UC_LSB;
+
+		val->val1 = charge / 1000000L;
+		val->val2 = charge % 1000000L;
+	} else {
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Fetch a sample from the INA700 sensor and store it in an internal
+ * driver buffer
+ * See sensor_sample_fetch in sensor.h for more details.
+ * @param dev Pointer to the sensor device
+ *
+ * @return 0 if successful, negative errno code if failure.
+ */
+static int ina700_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct ina700_data *data = dev->data;
+	const struct ina700_config *config = dev->config;
+	int ret;
+	enum sensor_channel_ina700 ina_chan = (enum sensor_channel_ina700)chan;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE &&
+	    chan != SENSOR_CHAN_DIE_TEMP && chan != SENSOR_CHAN_CURRENT &&
+	    chan != SENSOR_CHAN_POWER && ina_chan != SENSOR_CHAN_ENERGY &&
+	    ina_chan != SENSOR_CHAN_CHARGE) {
+		return -ENOTSUP;
+	}
+
+	if ((chan == SENSOR_CHAN_ALL) || (chan == SENSOR_CHAN_VOLTAGE)) {
+		ret = ina700_reg_read_16(&config->bus, INA700_REG_BUS_VOLT, &data->bus_voltage);
+		if (ret < 0) {
+			LOG_ERR("Failed to read bus voltage");
+			return ret;
+		}
+	}
+
+	if ((chan == SENSOR_CHAN_ALL) || (chan == SENSOR_CHAN_DIE_TEMP)) {
+		ret = ina700_reg_read_16(&config->bus, INA700_REG_TEMPERATURE, &data->temperature);
+		if (ret < 0) {
+			LOG_ERR("Failed to read temperature");
+			return ret;
+		}
+	}
+
+	if ((chan == SENSOR_CHAN_ALL) || (chan == SENSOR_CHAN_CURRENT)) {
+		ret = ina700_reg_read_16(&config->bus, INA700_REG_CURRENT, &data->current);
+		if (ret < 0) {
+			LOG_ERR("Failed to read current");
+			return ret;
+		}
+	}
+
+	if ((chan == SENSOR_CHAN_ALL) || (chan == SENSOR_CHAN_POWER)) {
+		ret = ina700_reg_read_24(&config->bus, INA700_REG_POWER, &data->power);
+		if (ret < 0) {
+			LOG_ERR("Failed to read power");
+			return ret;
+		}
+	}
+
+	if ((chan == SENSOR_CHAN_ALL) || (ina_chan == SENSOR_CHAN_ENERGY)) {
+		ret = ina700_reg_read_40(&config->bus, INA700_REG_ENERGY, &data->energy);
+		if (ret < 0) {
+			LOG_ERR("Failed to read energy");
+			return ret;
+		}
+	}
+
+	if ((chan == SENSOR_CHAN_ALL) || (ina_chan == SENSOR_CHAN_CHARGE)) {
+		ret = ina700_reg_read_40(&config->bus, INA700_REG_CHARGE, &data->charge);
+		if (ret < 0) {
+			LOG_ERR("Failed to read charge");
+			return ret;
```

Comment:
`INA700_REG_BUS_VOLT` through `INA700_REG_CHARGE` are sequential 0x5 - 0xa, it's much faster to just read all of it as a single i2c burst read.